File: app/src/main/java/com/amusoft/gdgfirechat/MainActivity.java
package com.amusoft.gdgfirechat;

import android.app.Activity;
import android.app.NotificationManager;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.database.DataSetObserver;
import android.net.Uri;
import android.os.Bundle;
import android.view.KeyEvent;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewConfiguration;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ListView;
import android.widget.Toast;

import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.core.app.NotificationCompat;

import com.google.android.gms.tasks.OnSuccessListener;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;
import com.google.firebase.storage.FirebaseStorage;
import com.google.firebase.storage.OnProgressListener;
import com.google.firebase.storage.StorageReference;
import com.google.firebase.storage.UploadTask;

import java.lang.reflect.Field;
import java.util.Objects;

public class MainActivity extends AppCompatActivity {

    private static final int RC_TAKE_PICTURE = 101;
    private static final int PROGRESS_NOTIFICATION_ID = 2;
    FirebaseDatabase database = FirebaseDatabase.getInstance();
    FirebaseStorage storage = FirebaseStorage.getInstance();


    // Setup our Firebase mFirebaseRef
    DatabaseReference mFirebaseRef = database.getReference("chat");
    StorageReference mStorageReref = storage.getReference();


    private String mUsername;

    private ValueEventListener mConnectedListener;
    private FirebaseListAdapter mChatListAdapter;


    ListView listView;
    Button sendphoto;
    EditText inputText;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        // Make sure we have a mUsername
        setupUsername();
        setToolBar();
        getOverflowMenu();
//
//        setTitle("Chatting as " + mUsername);


        // Setup our input methods. Enter key on the keyboard or pushing the send button
        inputText = (EditText) findViewById(R.id.chat_editText);
        sendphoto = (Button) findViewById(R.id.sendphoto);
        inputText.setOnKeyListener(new View.OnKeyListener() {
            public boolean onKey(View v, int keyCode, KeyEvent event) {
                // If the event is a key-down event on the "enter" button
                if ((event.getAction() == KeyEvent.ACTION_DOWN) && (keyCode == KeyEvent.KEYCODE_ENTER)) {
                    // Perform action on key press
                    final String question = inputText.getText().toString();
                    sendMessage(setupUsername(), question);
                    inputText.setText("");


                    return true;
                }

                return false;
            }
        });
        sendphoto.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                sendPhoto();
            }
        });

    }

    private void sendPhoto() {

        // Pick an image from storage
        Intent intent = new Intent(Intent.ACTION_GET_CONTENT);
        intent.setType("image/*");
        startActivityForResult(intent, RC_TAKE_PICTURE);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (requestCode == RC_TAKE_PICTURE) {
            if (resultCode == Activity.RESULT_OK) {

                if (data != null) {


                    Uri mFileUri = data.getData();

                    System.out.print(mFileUri);
                    uploadFromUri(mFileUri);
                } else {

                }
            } else {
                Toast.makeText(this, "Taking picture failed.", Toast.LENGTH_SHORT).show();
            }
        }
    }

    private void uploadFromUri(Uri mFileUri) {

        // [START get_child_ref]
        // Get a reference to store file at photos/<FILENAME>.jpg
        final StorageReference photoRef = mStorageReref.child("photos")
                .child(mFileUri.getLastPathSegment());
        // [END get_child_ref]


        // Upload file to Firebase Storage
        photoRef.putFile(mFileUri).addOnSuccessListener(new OnSuccessListener<UploadTask.TaskSnapshot>() {
            @Override
            public void onSuccess(UploadTask.TaskSnapshot taskSnapshot) {

                // [START_EXCLUDE
                String imageuri = Objects.requireNonNull(taskSnapshot.getMetadata()).getContentDisposition();
                sendMessage(setupUsername(), imageuri);
                dismissProgressNotification();

                // [END_EXCLUDE]

            }
        }).addOnProgressListener(new OnProgressListener<UploadTask.TaskSnapshot>() {
            @Override
            public void onProgress(UploadTask.TaskSnapshot taskSnapshot) {
                showProgressNotification(getString(R.string.progress_uploading),
                        taskSnapshot.getBytesTransferred(),
                        taskSnapshot.getTotalByteCount());
            }
        })
        ;


    }

    @Override
    public void onStart() {
        super.onStart();

        // Setup our view and list adapter. Ensure it scrolls to the bottom as data changes

        // Tell our list adapter that we only want 50 messages at a time
        mChatListAdapter = new FirebaseListAdapter(mFirebaseRef, this, mUsername) {
            @Override
            public void cleanup() {
                super.cleanup();
            }
        };
        listView = (ListView) findViewById(R.id.chat_listView);
        listView.setAdapter(mChatListAdapter);
        mChatListAdapter.registerDataSetObserver(new DataSetObserver() {
            @Override
            public void onChanged() {
                super.onChanged();
                listView.setSelection(mChatListAdapter.getCount() - 1);
            }
        });
        // Finally, a little indication of connection status
        mConnectedListener = mFirebaseRef.getRoot().child(".info/connected").addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(DataSnapshot dataSnapshot) {
                boolean connected = (Boolean) dataSnapshot.getValue();
                if (connected) {
                    Toast.makeText(getApplicationContext(), "Connected to Firebase", Toast.LENGTH_SHORT).show();
                } else {
                    Toast.makeText(getApplicationContext(), "Disconnected from Firebase", Toast.LENGTH_SHORT).show();
                }
            }

            @Override
            public void onCancelled(DatabaseError databaseError) {

            }

        });
    }

    @Override
    public void onStop() {
        super.onStop();
        mFirebaseRef.getRoot().child(".info/connected").removeEventListener(mConnectedListener);
        mChatListAdapter.cleanup();
    }

    private String setupUsername() {
        SharedPreferences prefs = getApplication().getSharedPreferences("ChatPrefs", 0);
        mUsername = prefs.getString("username", null);

        return mUsername;

    }

    private void sendMessage(String UserName, String question) {

        if (!question.equals("")) {
            // Create our 'model', a Chat object
            ChatMessage chat = new ChatMessage(question, UserName);
            // Create a new, auto-generated child of that chat location, and save our chat data there
            mFirebaseRef = database.getReference("chat");
            mFirebaseRef.push().setValue(chat);
            inputText.setText("");
        }
    }

    private void setToolBar() {
        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbarAdminDetails);
        if (toolbar != null) {

            setSupportActionBar(toolbar);
            setUpActionbar();
            getOverflowMenu();
            toolbar.setBackgroundColor(getResources().getColor(R.color.colorPrimary));
            toolbar.setTitleTextColor(getResources().getColor(R.color.white_pure));
        }
    }


    private void setUpActionbar() {
        if (getSupportActionBar() != null) {
            ActionBar bar = getSupportActionBar();
            bar.setTitle(getResources().getString(R.string.app_name));
            bar.setHomeButtonEnabled(false);
            bar.setDisplayShowHomeEnabled(false);
            bar.setDisplayHomeAsUpEnabled(false);
            bar.setDisplayShowTitleEnabled(true);
            bar.setNavigationMode(ActionBar.NAVIGATION_MODE_STANDARD);
        }


    }

    private void getOverflowMenu() {

        try {
            ViewConfiguration config = ViewConfiguration.get(this);

            Field menuKeyField = ViewConfiguration.class.getDeclaredField("sHasPermanentMenuKey");

            if (menuKeyField != null) {

                menuKeyField.setAccessible(true);
                menuKeyField.setBoolean(config, false);
                menuKeyField.isSynthetic();

            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_main_activty, menu);
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();
        if (id == R.id.action_log_out) {
            FirebaseAuth.getInstance().signOut();
            Intent i = new Intent(getApplicationContext(), LogInActivity.class);
            startActivity(i);
            finish();


        }
        return super.onOptionsItemSelected(item);

    }


    /**
     * Show notification with a progress bar.
     */
    protected void showProgressNotification(String caption, long completedUnits, long totalUnits) {
        int percentComplete = 0;
        if (totalUnits > 0) {
            percentComplete = (int) (100 * completedUnits / totalUnits);
        }

        NotificationCompat.Builder builder = (NotificationCompat.Builder) new NotificationCompat.Builder(this)
                .setSmallIcon(R.mipmap.ic_launcher)
                .setContentTitle(getString(R.string.app_name))
                .setContentText(caption)
                .setProgress(100, percentComplete, false)
                .setOngoing(true)
                .setAutoCancel(false);

        NotificationManager manager =
                (NotificationManager) getSystemService(Context.NOTIFICATION_SERVICE);

        manager.notify(PROGRESS_NOTIFICATION_ID, builder.build());
    }

    /**
     * Dismiss the progress notification.
     */
    protected void dismissProgressNotification() {
        NotificationManager manager =
                (NotificationManager) getSystemService(Context.NOTIFICATION_SERVICE);

        manager.cancel(PROGRESS_NOTIFICATION_ID);
    }


}

File: app/src/main/java/com/amusoft/gdgfirechat/ui/MainActivity.kt
package com.amusoft.gdgfirechat.ui

import android.annotation.SuppressLint
import android.app.Activity
import android.app.NotificationManager
import android.content.Context
import android.content.Intent
import android.database.DataSetObserver
import android.net.Uri
import android.os.Bundle
import android.util.Log
import android.view.*
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.NotificationCompat
import androidx.core.content.res.ResourcesCompat
import com.amusoft.gdgfirechat.R
import com.amusoft.gdgfirechat.adapter.FirebaseListAdapter
import com.amusoft.gdgfirechat.model.ChatMessage
import com.amusoft.gdgfirechat.showSnackBar
import com.amusoft.gdgfirechat.stripUsername
import com.amusoft.gdgfirechat.toUrlefy
import com.android.volley.DefaultRetryPolicy
import com.android.volley.Request
import com.android.volley.RequestQueue
import com.android.volley.toolbox.BasicNetwork
import com.android.volley.toolbox.DiskBasedCache
import com.android.volley.toolbox.HurlStack
import com.android.volley.toolbox.JsonObjectRequest
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.*
import com.google.firebase.storage.FirebaseStorage
import com.google.firebase.storage.StorageReference
import kotlinx.android.synthetic.main.activity_main.*
import org.json.JSONObject

class MainActivity : AppCompatActivity() {

    private val database: FirebaseDatabase = FirebaseDatabase.getInstance()

    private val storage: FirebaseStorage = FirebaseStorage.getInstance()

    // Setup our Firebase firebaseRef
    private val firebaseRef: DatabaseReference = database.getReference("chat")

    private val storageReference: StorageReference = storage.reference


    private var username: String? = null

    private var connectedListener: ValueEventListener? = null
    private var chatListAdapter: FirebaseListAdapter? = null

    private val manager by lazy { getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        setupUsername()
        setToolBar()
        getOverflowMenu()

        chat_editText.setOnKeyListener(View.OnKeyListener { _, keyCode, event ->
            // If the event is a key-down event on the "enter" button
            if (event.action == KeyEvent.ACTION_DOWN && keyCode == KeyEvent.KEYCODE_ENTER) {
                // Perform action on key press
                val question = chat_editText.text.toString()
                if(question.contains("@SophieBot")){
                    getAnswer("SophieBot", stripUsername(question))
                }

                sendMessage(setupUsername(), question)

                chat_editText.setText("")
                return@OnKeyListener true
            }

            false
        })

        sendphoto.setOnClickListener { sendPhoto() }

    }

    private fun sendPhoto() {
        // Pick an image from storage
        val intent = Intent(Intent.ACTION_GET_CONTENT)
        intent.type = "image/*"
        startActivityForResult(intent, RC_TAKE_PICTURE)
    }

    @SuppressLint("MissingSuperCall")
    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (requestCode == RC_TAKE_PICTURE) {
            if (resultCode == Activity.RESULT_OK && data != null) {
                val mFileUri = data.data
                print(mFileUri)
                uploadFromUri(mFileUri!!)
            } else {
                Log.d("Image selection", "Image not retrieved")
            }
        } else {
            Toast.makeText(this, "Taking picture failed.", Toast.LENGTH_SHORT).show()
        }
    }

    private fun uploadFromUri(fileUri: Uri) {

        // [START get_child_ref]
        // Get a reference to store file at photos/<FILENAME>.jpg
        val photoRef = storageReference.child("photos")
                .child(fileUri.lastPathSegment!!)
        // [END get_child_ref]

        // Upload file to Firebase Storage
        photoRef.putFile(fileUri).addOnSuccessListener { taskSnapshot ->
            // [START_EXCLUDE
            val imageUri = taskSnapshot.metadata?.contentDisposition
            sendMessage(setupUsername(), imageUri!!)
            dismissProgressNotification()

            // [END_EXCLUDE]
        }.addOnProgressListener { taskSnapshot ->
            showProgressNotification(getString(R.string.progress_uploading),
                    taskSnapshot.bytesTransferred,
                    taskSnapshot.totalByteCount)
        }


    }

    public override fun onStart() {
        super.onStart()

        // Setup our view and list adapter. Ensure it scrolls to the bottom as data changes

        // Tell our list adapter that we only want 50 messages at a time
        chatListAdapter = object : FirebaseListAdapter(firebaseRef, this, username!!) {
            override fun cleanup() {
                super.cleanup()
            }
        }

        chat_listView.adapter = chatListAdapter

        chatListAdapter!!.registerDataSetObserver(object : DataSetObserver() {
            override fun onChanged() {
                super.onChanged()
                chat_listView.setSelection(chatListAdapter!!.count - 1)
            }
        })
        // Finally, a little indication of connection status
        connectedListener = firebaseRef.root.child(".info/connected")
                .addValueEventListener(object : ValueEventListener {
                    override fun onDataChange(dataSnapshot: DataSnapshot) {

                        val connected = (dataSnapshot.value as Boolean)

                        if (connected) {
                            Toast.makeText(applicationContext, "Connected to Firebase", Toast.LENGTH_SHORT).show()
                        } else {
                            Toast.makeText(applicationContext, "Disconnected from Firebase", Toast.LENGTH_SHORT).show()
                        }
                    }

                    override fun onCancelled(databaseError: DatabaseError) {

                    }

                })
    }

    public override fun onStop() {
        super.onStop()
        firebaseRef.root.child(".info/connected").removeEventListener(connectedListener!!)
        chatListAdapter!!.cleanup()
    }

    private fun setupUsername(): String {

        val prefs = application.getSharedPreferences("ChatPrefs", 0)

        username = prefs.getString("username", null)

        return username ?: ""

    }

    private fun sendMessage(userName: String, question: String) {

        if (question != "") {
            // Create our 'model', a Chat object
            val chat = ChatMessage(question, userName)
            // Create a new, auto-generated child of that chat location, and save our chat data there
            firebaseRef.push().setValue(chat)
            chat_editText.text.clear()
        }
    }
    private fun getAnswer(username: String, question: String) {
        var apilink ="https://publicapi.sophiebot.ai/answer/"
        var apireuest=apilink+ toUrlefy(question)
        //TODO send request to api and add it to firebase

        // Instantiate the cache
        val cache = DiskBasedCache(cacheDir, 1024 * 1024) // 1MB cap
        // Set up the network to use HttpURLConnection as the HTTP client.
        val network = BasicNetwork(HurlStack())
        // Instantiate the RequestQueue with the cache and network. Start the queue.
        val requestQueue = RequestQueue(cache, network).apply {
            start() }

        val jsonObjectRequest = JsonObjectRequest(
            Request.Method.GET, apireuest, null,
            { response ->
                var strResp = response.toString()
                val jsonObj: JSONObject = JSONObject(strResp)
                jsonObj.get("answer")?.let {
                    val chat = ChatMessage(username,it.toString())
                    // Create a new, auto-generated child of that chat location, and save our chat data there
                    sendMessage(chat.author,chat.message)
                }
            },
            { error ->
                // TODO: Handle error
                showSnackBar(chat_listView,"Disconnected from the internet, this version of SophieBot cannot work offline")
            }
        )
        jsonObjectRequest.retryPolicy = DefaultRetryPolicy(
            500000000,
            DefaultRetryPolicy.DEFAULT_MAX_RETRIES,
            DefaultRetryPolicy.DEFAULT_BACKOFF_MULT
        )

// Access the RequestQueue through your singleton class.
        requestQueue.add(jsonObjectRequest)


    }




    private fun setToolBar() {

        setSupportActionBar(toolbarAdminDetails)
        supportActionBar?.run {
            title = resources.getString(R.string.app_name)
            setHomeButtonEnabled(false)
            setDisplayShowHomeEnabled(false)
            setDisplayHomeAsUpEnabled(false)
            setDisplayShowTitleEnabled(true)
        }

        getOverflowMenu()

        toolbarAdminDetails.setBackgroundColor(ResourcesCompat.getColor(resources, R.color.colorPrimary, null))

        toolbarAdminDetails.setTitleTextColor(ResourcesCompat.getColor(resources, R.color.white_pure, null))
    }

    @SuppressLint("SoonBlockedPrivateApi")
    private fun getOverflowMenu() {

        try {
            val config = ViewConfiguration.get(this)

            val menuKeyField = ViewConfiguration::class.java.getDeclaredField("sHasPermanentMenuKey")

            menuKeyField.run {
                isAccessible = true
                setBoolean(config, false)
                isSynthetic
            }

        } catch (e: Exception) {
            e.printStackTrace()
        }

    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_main_activty, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {

        if (item.itemId == R.id.action_log_out) {

            FirebaseAuth.getInstance().signOut()

            startActivity(Intent(this, LogInActivity::class.java))
            finish()

            return true
        }

        return super.onOptionsItemSelected(item)

    }


    /**
     * Show notification with a progress bar.
     */
    protected fun showProgressNotification(caption: String, completedUnits: Long, totalUnits: Long) {

        var percentComplete = 0

        if (totalUnits > 0) {
            percentComplete = (100 * completedUnits / totalUnits).toInt()
        }

        val builder = NotificationCompat.Builder(this)
                .setSmallIcon(R.mipmap.ic_launcher)
                .setContentTitle(getString(R.string.app_name))
                .setContentText(caption)
                .setProgress(100, percentComplete, false)
                .setOngoing(true)
                .setAutoCancel(false) as NotificationCompat.Builder



        manager.notify(PROGRESS_NOTIFICATION_ID, builder.build())
    }

    /**
     * Dismiss the progress notification.
     */
    private fun dismissProgressNotification() {
        manager.cancel(PROGRESS_NOTIFICATION_ID)
    }

    companion object {
        private val RC_TAKE_PICTURE = 101
        private val PROGRESS_NOTIFICATION_ID = 2
    }


}
